clarity_ext/service/dilution/service.py

```diff
@@ -431,6 +440,54 @@ def __repr__(self):
             tuple(self.update_info))
 
 
+class VirtualTransferBatch(object):
+    """Contains the original "virtual" transfers we want performed. These will translate into
+    a series of actual SingleTransfer objects."""
+
+    def __init__(self, transfers):
+        """Creates a virtual batch and updates each transfer so it has a pointer to the it"""
+        self.virtual_transfers = VirtualTransferBatch.transfers_by_output(transfers)
+        self._transfers = transfers
+
+        for transfer in transfers:
+            transfer.virtual_batch = self
+
+    def __getitem__(self, transfer):
+        """Returns the virtual transfer by the transfer"""
+        return self.virtual_transfers[transfer.target_location.artifact.id]
+
+    @staticmethod
+    def transfers_by_output(transfers):
+        def group():
+            def group_key(transfer):
+                return transfer.target_location.artifact.id
+            s = sorted(transfers, key=group_key)
+            return {k: list(t) for k, t in groupby(s, key=group_key)}
+        return {k: VirtualTransfer(t) for k, t in group().items()}
+
+
+class VirtualTransfer(object):
```

Comment:
Maybe have a name that referers to pools?

Reply:
Yes, I think my tendency to try to be generic might be in the way here. I was though thinking that (further down the line) this would encapsulate the initial transfer that the RE does, even those that are not pools. I then go ahead and comment that it has to do with pools (only), which was a mistake. I hope I can clarify this in future comments.